proposed/runtime-binding.md

```diff
@@ -0,0 +1,349 @@
+
+Let's look at an example.
+
+Requested version:
+
+* `2.1.0`
+
+Installed versions:
+
+* `1.1.17`
+* `2.2.0`
+* `2.2.1`
+* `2.2.5`
+* `2.3.1`
+* `3.0.0`
+
+In this scenario, the `2.2.5` runtime will be selected.
+
+Let's look at another example.
+
+Requested version:
+
+* `2.1.0`
+
+Installed versions:
+
+* `1.1.17`
+* `3.0.0`
+
+In this scenario, no runtime will be selected and instead the host prints a helpful error message (similar to the one displayed in the **Specifing a .NET Core Runtime Version** section).
+
+We have found that this behavior is problematic for scenarios like [global tools](https://twitter.com/KathleenDollard/status/1079811275641696256) and expect similar challenges for client applications (new with 3.0). To mitigate these scenarios, we will expose a set of configuration knobs to enable developers to opt-into a specific roll-forward behaviors for their app or their environment. We will write guidance for those scenarios to guide developers and end-users to opt-in to roll-forward for scenarios that will benefit from it. We will not provide templates that set these knobs automatically for developers.
+
+## Major Version Selection -- Opt-in Behavior using Configuration Knobs
+
+> This runtime selection logic is used if an appropriate minor version for a given three-part runtime version cannot be found and an application or environment has been configured for major-version roll-forward via **Configuration Knobs** (defined in a following section).
+
+Given a request for `x.y.z` version, the host will look for an `x` runtime version, as described for **Minor Version Selection** above. If it cannot find an `x` runtime, then the host will look for all runtimes that are higher than `x`. It will select the lowest major version within that set and then select the lowest minor and highest patch version within that set. This behavior is oriented on a best effort attempt at selecting a compatible installed product version. That's why the lowest minor version is selected, not the highest one.
+
+**User Impact:** When a user runs an application on a machine that does not have `x` installed, but has a later runtime installed, then the app will transparently run on that version. The user does not need to configure their applications to use that later minor version. There is some risk in this scenario that the app will crash due to incompatibilities in the runtime relative to the runtime that the app targets and that it will be difficult to diagnose the cause (incompatibility)and the solution (installing a older runtime). If that app runs successfully, then the user will be happy. The tension between compatibility and deployment convenience is why this scenario is opt-in.
```

Comment:
I think you meant "applications to use that later **major** version."

Reply:
Will there be any info or warning text when the app starts to warn the user for the major version roll forward? 

I see a developer just assuming there app will work for all future versions without knowing that version 4.0 changed his assumptions and the app crashes. If the app **transparently** rolls forward and crashes the user never realizes they can solve the problem them selves by installing the appropriate version of .NET Core and instead blame the original developer.

Reply:
Our philosophy is that the developer owns stdout and stderr. In part, one app's stdout in another app's stdin. stderr is more an option, but is also not perfect.

My belief is that it should be uncommon for apps to run in environments where major version rollforward is enabled. If apps crash in such an environment, it should be possible to connect the opt-in rollforward setting with the crash and consider that the setting is the cause of the crash if no other more obvious causes are identified.

Fair?

Reply:
What other mechanism do you think should be used to notify about the roll-forward happening.
On Windows we were discussing that maybe we should write an information event into the event log.

Reply:
I like the idea of it getting put out as a runtime event.